### PR TITLE
provider/openstack: Security Group Rule fixes

### LINKED ISSUE #3788: openstack_compute_secgroup_v2 does not remove unused rules
If a rule is changed or removed from a `openstack_compute_secgroup_v2`, the stale rule is not removed from openstack.  Eg.,

```
$ openstack security group rule list 86bb9323-0543-491b-8594-ef0459bc05ff
+--------------------------------------+-------------+-----------+------------+
| ID                                   | IP Protocol | IP Range  | Port Range |
+--------------------------------------+-------------+-----------+------------+
| 14be9699-f20d-4711-bea4-9196d4f5f59d | tcp         | 0.0.0.0/0 | 80:80      |
+--------------------------------------+-------------+-----------+------------+
```

Changing the a rule for tcp/80 -> tcp/2375:
```
resource "openstack_compute_secgroup_v2" "secgroup_1" {
  name = "my_secgroup"
  description = "my security group"
  rule {
    from_port = 2375 # was 80
    to_port = 2375 # was 80
    ip_protocol = "tcp"
    cidr = "0.0.0.0/0"
  }
}
```

```
$ ./bin/terraform apply
openstack_compute_secgroup_v2.secgroup_1: Refreshing state... (ID: 86bb9323-0543-491b-8594-ef0459bc05ff)
openstack_networking_network_v2.network_1: Refreshing state... (ID: 8c29e770-c10f-4d4e-8cd0-2ea80ca0d211)
openstack_networking_router_v2.router_1: Refreshing state... (ID: 91081b73-abf9-4fe5-887c-61a3c78a111a)
openstack_networking_subnet_v2.subnet_1: Refreshing state... (ID: 7b217b45-88cc-4584-a6ee-c999a8ac81be)
openstack_networking_router_interface_v2.router_interface_1: Refreshing state... (ID: 48ce015e-e227-499c-90fe-4889a9421aee)
openstack_networking_router_v2.router_1: Modifying...
  admin_state_up: "1" => "0"
openstack_compute_secgroup_v2.secgroup_1: Modifying...
  rule.0.from_port: "80" => "2375"
  rule.0.to_port:   "80" => "2375"
openstack_networking_router_v2.router_1: Modifications complete
openstack_compute_secgroup_v2.secgroup_1: Modifications complete

Apply complete! Resources: 0 added, 2 changed, 0 destroyed.

The state of your infrastructure has been saved to the path
below. This state is required to modify and destroy your
infrastructure, so keep it safe. To inspect the complete state
use the `terraform show` command.

State path: terraform.tfstate
```

```
$ openstack security group rule list 86bb9323-0543-491b-8594-ef0459bc05ff
+--------------------------------------+-------------+-----------+------------+
| ID                                   | IP Protocol | IP Range  | Port Range |
+--------------------------------------+-------------+-----------+------------+
| 14be9699-f20d-4711-bea4-9196d4f5f59d | tcp         | 0.0.0.0/0 | 80:80      |
| 7a4a4381-3f1f-4eeb-8839-addae5e1317d | tcp         | 0.0.0.0/0 | 2375:2375  |
+--------------------------------------+-------------+-----------+------------+
```

### LINKED ISSUE #3816: provider/openstack: updating security groups is broken
I found two issues:

1. Existing security group despite not being changed at all in template is marked as changed on every `terraform apply`
2. Because of it, every instance associated with this security group _deletes the association and then readds it_. I am convinced that is is absolutely not required to re-associate security group after rules have changed. It actually can lead to downtime of the instance because if gap when SG is not yet associated

### PR DESCRIPTION
**Pull request by jtopjian**
This commit fixes an issue with security group rules where the rules
were not being correctly computed due to a typo in the rule map.

Once rules were successfully computed, the rules then needed to be
converted into a Set so they can be correctly ordered.

Fixes #3788 
Fixes #3816

### CODE REVIEW

**Comment by jtopjian** [comment]
@Fodoj @jhoblitt Let me know if this fixes your issues.

**Comment by jtopjian** [comment]
@phinze This patch is doing a similar `TypeList` to `TypeSet` conversion as #3651. Does this need a migration or will the data type conversion be done implicitly again?

File: builtin/providers/openstack/resource_openstack_compute_secgroup_v2.go [hunk 1]
```diff
@@ -283,12 +283,12 @@ func rulesToMap(sgrs []secgroups.Rule) []map[string]interface{} {
...
-			"to_port":     sgr.ToPort,
-			"ip_protocol": sgr.IPProtocol,
-			"cidr":        sgr.IPRange.CIDR,
-			"group":       sgr.Group.Name,
+			"id":            sgr.ID,
+			"from_port":     sgr.FromPort,
+			"to_port":       sgr.ToPort,
+			"ip_protocol":   sgr.IPProtocol,
+			"cidr":          sgr.IPRange.CIDR,
+			"from_group_id": sgr.Group.Name,
```
Comment (phinze) [inline builtin/providers/openstack/resource_openstack_compute_secgroup_v2.go]: Is this change exercised in any of the tests?

**Comment by phinze** [comment]
@jtopjian this looks to be the same scenario so I think we're okay to skip the state migration. Code LGTM with one inline question. :ok_hand:

File: builtin/providers/openstack/resource_openstack_compute_secgroup_v2.go (hunk 1 above)
Comment (jtopjian) [inline builtin/providers/openstack/resource_openstack_compute_secgroup_v2.go]: Ah - it's not. Thanks!

Comment (jtopjian) [inline builtin/providers/openstack/resource_openstack_compute_secgroup_v2.go]: Well, that escalated quickly!

**Comment by Fodoj** [comment]
There seems to be some issue with (I didn't test with last commit though): terraform tries to delete SG from the instance, deletes it (it's no associated any more) but fails and shows "Faield to remove". But the weird part is that it happens during `terraform apply`, meaning that it actually should not even try to remove SG from the instance.

**Comment by Fodoj** [comment]
Ignore last comment, it seems to be no related to this PR.

**Comment by Fodoj** [comment]
More info:

1. AKAIK security group _names_ are not unique in Openstack => using names for SGs list of instance is wrong then
2. It one passes ID to security_groups attribute of an instance instead of name then weird bugs that the one I described above will happen:

Here it has name in state file, id passed in and then it explodes.

```
2015/11/10 16:52:04 [DEBUG] terraform-provider-openstack: 2015/11/10 16:52:04 [DEBUG] oldSGSet: &{0x575320 map[47827970:security_group_1] {{0 0} 1}}
2015/11/10 16:52:04 [DEBUG] terraform-provider-openstack: 2015/11/10 16:52:04 [DEBUG] newSGSet: &{0x575320 map[1589016600:84c4b52f-e023-40d6-9edf-7dbd73b91d7c] {{0 0} 1}}
````

**Comment by jtopjian** [comment]
Yeah, the places where to specify a security group name and ID are very inconsistent in OpenStack.

The Nova instances expect security group names. Neutron Ports expect IDs. The Nova security group rules expect an ID, but will return a name and not the ID.

Unfortunately there's not a lot that can be done here besides documentation. I tried doing some workarounds to enable security group IDs in the `security_groups` attribute of instances in #2009, but ultimately wasn't successful :(

**Comment by jtopjian** [comment]
@Fodoj With the wonkiness of security groups aside, does this patch work for you? 

@phinze If you have a minute, could you do a quick code review? Adding the missing test you caught required modification to a few other areas (again, thanks for catching it :)

**Comment by Fodoj** [comment]
Yes, after using names instead of IDs everything started working as expected



Best regards,
Kirill Shirinkin

**Comment by phinze** [comment]
@jtopjian This is looking good! Merge at will. :+1:

**Comment by jhoblitt** [comment]
@jtopjian Sorry for the slow response... It has been one of "those weeks". I will test and report back later this morning but please don't hold up merging on my account.

**Comment by jtopjian** [comment]
@phinze @Fodoj Thank you both!

@jhoblitt No problem at all. I'll merge this later today. If you don't get a chance to test before then, please let me know if you run into any issues afterward.

**Pull request merged**